builder/cache/cache.go

```diff
@@ -0,0 +1,152 @@
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package cache
+
+import (
+	"container/list"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"upper.io/db.v2/builder/cache/hashstructure"
+)
+
+const defaultCapacity = 128
+
+// Cache holds a map of volatile key -> values.
+type Cache struct {
+	cache map[string]*list.Element
+	li    *list.List
+	cap   int
+	mu    sync.RWMutex
+}
+
+type item struct {
+	key   string
+	value interface{}
+}
+
+// NewCacheWithCapacity initializes a new caching space with the given
+// capacity.
+func NewCacheWithCapacity(cap int) (*Cache, error) {
+	if cap < 1 {
+		return nil, errors.New("Capacity must be greater than zero.")
+	}
+	return &Cache{
+		cache: make(map[string]*list.Element),
+		li:    list.New(),
+		cap:   cap,
+	}, nil
+}
+
+// NewCache initializes a new caching space with default settings.
+func NewCache() *Cache {
+	c, err := NewCacheWithCapacity(defaultCapacity)
+	if err != nil {
+		panic(err.Error()) // Should never happen as we're not providing a negative defaultCapacity.
+	}
+	return c
+}
+
+// Read attempts to retrieve a cached value as a string, if the value does not
+// exists returns an empty string and false.
+func (c *Cache) Read(h Hashable) (string, bool) {
+	if v, ok := c.ReadRaw(h); ok {
+		if s, ok := v.(string); ok {
+			return s, true
+		}
+	}
+	return "", false
+}
+
+// ReadRaw attempts to retrieve a cached value as an interface{}, if the value
+// does not exists returns nil and false.
+func (c *Cache) ReadRaw(h Hashable) (interface{}, bool) {
+	c.mu.RLock()
+	data, ok := c.cache[h.Hash()]
+	c.mu.RUnlock()
+	if ok {
+		return data.Value.(*item).value, true
+	}
+	return nil, false
+}
+
+// Write stores a value in memory. If the value already exists its overwritten.
+func (c *Cache) Write(h Hashable, value interface{}) {
+	key := h.Hash()
+
+	c.mu.Lock()
+
+	if el, ok := c.cache[key]; ok {
+		el.Value.(*item).value = value
+		c.li.MoveToFront(el)
+		c.mu.Unlock()
+		return
+	}
+
+	c.cache[key] = c.li.PushFront(&item{key, value})
+
+	if c.li.Len() > c.cap {
+		el := c.li.Remove(c.li.Back())
+		delete(c.cache, el.(*item).key)
+		if p, ok := el.(*item).value.(HasOnPurge); ok {
+			p.OnPurge()
+		}
+	}
+
+	c.mu.Unlock()
```

Comment:
I ran the benchmarks and there's not really much speed gain for not using `defer` in this case, so I am adding `defer` here and on `Clear`.